Supports manager create

kbrock opened the pull request:
Depends: 

- [x] https://github.com/ManageIQ/manageiq/pull/21553

### Goal

Moving from a custom method `supported_for_create?` to `supports :create`
Moving from parent classes defining `:create` to the actual classes themselves defining it.

Also converted network managers and storage across to follow the same pattern.

only real question marks I had was physical infra provider. I did that. figure it would be easier to find one extra one that I added rather than missing one.

I also noticed ui classic had 2 mixins that would be nice to remove. But there are quite a few other classes that follow this code path and it seemed safer to leave that code be. Not sure how many of these classes do declare `supports :update`

### Other required PRs

- [ ] https://github.com/ManageIQ/manageiq-ui-classic/pull/7985
- [x] https://github.com/ManageIQ/manageiq-providers-amazon/pull/731
- [x] https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/270
- [x] https://github.com/ManageIQ/manageiq-providers-autosde/pull/120
- [x] https://github.com/ManageIQ/manageiq-providers-autosde/pull/110
- [x] https://github.com/ManageIQ/manageiq-providers-azure/pull/486
- [x] https://github.com/ManageIQ/manageiq-providers-azure_stack/pull/71
- [x] https://github.com/ManageIQ/manageiq-providers-foreman/pull/94
- [x] https://github.com/ManageIQ/manageiq-providers-google/pull/206
- [x] https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/306
- [x] https://github.com/ManageIQ/manageiq-providers-ibm_power_vc/pull/33
- [x] https://github.com/ManageIQ/manageiq-providers-ibm_terraform/pull/71
- [x] https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/454
- [x] ~~https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/192~~
- [x] https://github.com/ManageIQ/manageiq-providers-lenovo/pull/343
- [x] https://github.com/ManageIQ/manageiq-providers-nsxt/pull/50
- [x] https://github.com/ManageIQ/manageiq-providers-nuage/pull/257
- [x] https://github.com/ManageIQ/manageiq-providers-openshift/pull/217
- [x] https://github.com/ManageIQ/manageiq-providers-openstack/pull/757
- [ ] https://github.com/ManageIQ/manageiq-providers-oracle_cloud/pull/37
- [x] https://github.com/ManageIQ/manageiq-providers-ovirt/pull/587
- [x] https://github.com/ManageIQ/manageiq-providers-redfish/pull/149
- [x] https://github.com/ManageIQ/manageiq-providers-scvmm/pull/197
- [x] https://github.com/ManageIQ/manageiq-providers-vmware/pull/767

agrare commented on app/models/manageiq/providers/network_manager.rb:
```diff
@@ -9,6 +9,7 @@ class << model_name
     end
 
+    supports_not :create
     supports_not :ems_network_new
```
Actually we could probably drop `ems_network_new` also here, looks like it is used by the UI to check if edit should be allowed but we could simply use `supports?(:update)` instead

kbrock replied:
lol. do we have an emoji for scope creep?

I'll create a separate PR. it will conflict with this PR but it looks straight forward enough.
I do want to move these over to CRUD, but right now, my goal is getting us to `supports` (in any form)

agrare replied:
Yeah definitely you can do that in a different PR, just caught my eye here since `supports :create` and `supports :ems_network_new` ostensibly track the same feature.

agrare replied:
That was my first impression but now seeing how it is used (per https://github.com/ManageIQ/manageiq/pull/21573/files#r756180425) I think `s/ems_network_new/update`

agrare replied:
Correct, number B :wink:

agrare commented:
@miq-bot cross-repo-tests manageiq-ui-classic, manageiq-api, https://github.com/ManageIQ/manageiq-providers-autosde/pull/110, https://github.com/ManageIQ/manageiq-providers-foreman/pull/94, https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/306, https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/454, https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/192, https://github.com/ManageIQ/manageiq-providers-nsxt/pull/50, https://github.com/ManageIQ/manageiq-providers-nuage/pull/257, https://github.com/ManageIQ/manageiq-providers-openshift/pull/217

agrare commented:
@kbrock looks good to me but there are some related test failures in the manageiq-api repo: https://app.travis-ci.com/github/ManageIQ/manageiq-cross_repo-tests/jobs/550451415#L2474-L2490 (the authentication_spec failures are sporadic but the providers_spec ones are not)

kbrock commented:
pushed:
- added the `supports :update`
- moved supports update away from the base ems provider and to the base network provider. please verify this choice
- added more repos to the test, will need to expand the cross repo test
- Having trouble running specs locally due to ovirt changes. Will fix the supports issue once that is fixed

kbrock commented:
wip: we are going to remove entries from the hierarchy and move all the supports create/update to the leaf nodes that need them

agrare commented on app/models/ext_management_system.rb:
```diff
@@ -41,11 +41,7 @@ def self.api_allowed_attributes
-
-  def self.supported_for_create?
-    !reflections.include?("parent_manager")
+    subclasses_supporting(:create)
```
:heart:

agrare commented:
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-ui-classic/pull/7985, manageiq-api, https://github.com/ManageIQ/manageiq-providers-amazon/pull/731, https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/270, https://github.com/ManageIQ/manageiq-providers-autosde/pull/120, https://github.com/ManageIQ/manageiq-providers-azure/pull/486, https://github.com/ManageIQ/manageiq-providers-azure_stack/pull/71, https://github.com/ManageIQ/manageiq-providers-foreman/pull/94, https://github.com/ManageIQ/manageiq-providers-google/pull/206, https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/306, https://github.com/ManageIQ/manageiq-providers-ibm_power_vc/pull/33, https://github.com/ManageIQ/manageiq-providers-ibm_terraform/pull/71, https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/454, https://github.com/ManageIQ/manageiq-providers-lenovo/pull/343, https://github.com/ManageIQ/manageiq-providers-nsxt/pull/50, https://github.com/ManageIQ/manageiq-providers-nuage/pull/257, https://github.com/ManageIQ/manageiq-providers-openshift/pull/217, https://github.com/ManageIQ/manageiq-providers-openstack/pull/757, https://github.com/ManageIQ/manageiq-providers-oracle_cloud/pull/37, https://github.com/ManageIQ/manageiq-providers-ovirt/pull/587, https://github.com/ManageIQ/manageiq-providers-redfish/pull/149, https://github.com/ManageIQ/manageiq-providers-scvmm/pull/197, https://github.com/ManageIQ/manageiq-providers-vmware/pull/767

agrare commented:
There are a couple of stragglers but want to get this in
